funflow-checkpoints/src/Control/FunFlow/Checkpoints.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeOperators         #-}
+module Control.FunFlow.Checkpoints
+  ( CheckpointException(..)
+  , CheckpointT
+  , checkpoint
+  , runToCompletion
+  , runToCheckpoint
+  , runToCheckpointEither
+  ) where
+
+import           Control.Arrow
+import           Control.Arrow.Free
+import           Control.Category
+import           Control.Exception.Base (Exception, throw)
+import           Control.FunFlow
+import           Data.Either            (Either (..))
+import           Data.Eq                (Eq, (==))
+import           Data.Function          (flip, ($))
+import           Data.Maybe             (Maybe (..))
+import           Data.Proxy
+import           Data.String            (String)
+import           Data.Tuple             (fst, snd)
+import           Data.Typeable
+import           Text.Show              (Show)
+
+data CheckpointException =
+    -- | A checkpoint has been defined multiple times in the flow
+    RepeatedCheckpoint String
+    -- | A conditional branch has a checkpoint in.
+  | CheckpointInConditional String
+    -- | A branch responsible for catching errors has a checkpoint in.
+  | CheckpointInCatch String
+    -- | No matching checkpoint has been defined in the flow
+  | NoCheckpoint String
+  deriving (Eq, Show)
+
+instance Exception CheckpointException
+
+data CheckpointRef a = CheckpointRef (Proxy a) String
+  deriving Eq
+
+-- | Get the name of a constructed checkpoint
+checkpointName :: CheckpointRef a -> String
+checkpointName (CheckpointRef _ x) = x
+
+-- | A checkpoint arrow transfomer.
+data CheckpointT arr (eff :: * -> * -> *) ex a b where
+  Checkpoint :: Typeable a => CheckpointRef a -> CheckpointT arr eff ex a a
+  Unchecked :: arr eff ex a b -> CheckpointT arr eff ex a b
+  Seq :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex a c
+  Par :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex c d -> CheckpointT arr eff ex (a,c) (b,d)
+  Fanin :: CheckpointT arr eff ex a c -> CheckpointT arr eff ex b c -> CheckpointT arr eff ex (Either a b) c
+  Catch :: CheckpointT arr eff ex a b -> CheckpointT arr eff ex (a,ex) b -> CheckpointT arr eff ex a b
+
+instance Category (arr eff ex) => Category (CheckpointT arr eff ex) where
+  id = Unchecked id
+  (.) = flip Seq
+instance Arrow (arr eff ex) => Arrow (CheckpointT arr eff ex) where
+  arr f = Unchecked $ arr f
+  first f = Par f id
+  second = Par id
+  (***) = Par
+instance ArrowChoice (arr eff ex) => ArrowChoice (CheckpointT arr eff ex ) where
```

Comment:
You don't use the `ArrowChoice (arr eff ex)` constraint, as far as I can tell. But I'm guessing  that you need the constraint in order to interpret the syntactic `CheckpointT` (which, I see, is abstract), _i.e_ the constraint is part of the abstraction.

Worth documenting this.